docs/Debugging_Bono_Sprout_and_Homestead_with_GDB_and_Valgrind.md

```diff
@@ -1,35 +1,41 @@
+* Valgrind's memory access tracking hooks into malloc and free.  Unfortunately, pjsip uses its own memory management functions, and so mallocs/frees relatively rarely.  To disable this, modify `pjlib/src/pj/pool_caching.c`'s `pj_caching_pool_init` function to always set cp->max_capacity to 0.
 
--  make sure valgrind is installed on your system and you have the appropriate debug packages installed (`sudo apt-get install valgrind` and `sudo apt-get install sprout-dbg`)
+* Valgrind's thread safety tool (helgrind) tracks the order in which locks are taken, and reports on any lock cycles (which can in theory cause deadlocks). One of these locks generates a lot of benign results. To prevent these edit `pjsip\include\pjsip\sip_config.h` and set the value of `PJSIP_SAFE_MODULE` to 0.
 
--  disable monitoring of sprout (`sudo monit unmonitor -g sprout`)
+* Valgrind's thread safety tool also spots variables that are accessed on mutliple threads without locking the locking necessary to prevent race conditions. Pjsip's memory pools maintain some statistics that are not used by Clearwater, but that trip valgrind's race condition detection. To suppress this edit `pjlib/src/pj/pool_caching.c` and remove the bodies of the `cpool_on_block_alloc` and `cpool_on_block_free` (keeping only the bodies).
 
--  use ps to find the command line you are using to run Sprout (`ps -eaf | grep sprout`)
+To run Bono, Sprout or Homestead under valgrind (the example commands assume you are running sprout), the easiest way is to:
 
--  change to the sprout user (`sudo -u sprout bash`)
+-  Find the command line you are using to run Sprout (`ps -eaf | grep sprout`)
+-  Make sure valgrind is installed on your system and you have the appropriate debug packages installed (`sudo apt-get install valgrind` and `sudo apt-get install sprout-dbg`)
+-  Disable monitoring of sprout (`sudo monit unmonitor -g sprout`)
+-  Stop sprout (`sudo service sprout stop`)
+-  Change to the sprout user (`sudo -u sprout bash`)
+-  Change to a the /etc/clearwater directory
+-  Set up the library path (`export LD_LIBRARY_PATH=/usr/share/clearwater/sprout/lib:$LD_LIBRARY_PATH`)
+-  Run the executable under valgrind, enabling the appropriate valgrind options - for example, to use massif to monitor the Sprout heap `valgrind --tools=massif --massif-out-file=/var/log/sprout/massif.out.%p /usr/share/clearwater/bin/sprout <parameters>` (the --massif-out-file option is required to ensure the output is written to a directory where the sprout user has write permission). If any of Sprout parameters include a semi-colon, you must prefix this with a backslash otherwise the bash interpreter will interpret this as the end of the command.
 
--  change to a the /etc/clearwater directory
+You may find that after a while sprout reports that it is unable to create new file descriptors. If this happens, you need to run sprout as a service instead.
```

Comment:
I couldn't get that to work despite spending a couple of hours on it. I think you can only change your ulimit as root, but running sprout as root caused some other problems (though I can't remember the details).